README.md
[![Build Status](https://travis-ci.org/bitpanda-labs/loggo.svg?branch=master)](https://travis-ci.org/bitpanda-labs/loggo)
[![codecov.io](https://codecov.io/gh/bitpanda-labs/loggo/branch/master/graph/badge.svg)](https://codecov.io/gh/bitpanda-labs/loggo)
[![PyPI version](https://badge.fury.io/py/loggo.svg)](https://badge.fury.io/py/loggo)
[![Code style: black](https://img.shields.io/badge/code%20style-black-000000.svg)](https://github.com/psf/black)

# `@loggo`: automated logging for Python 3

<!--- Don't edit the version line below manually. Let bump2version do it for you. -->
> Version 10.1.3

> You find Python's builtin `logging` module repetitive, tedious and ugly, and the logs you do write with it clash with your otherwise awesome style. `loggo2` is here to help: it automates the boring stuff, simplifies the tricky stuff, hooks up effortlessly to [graylog](https://www.graylog.org/), and keeps an eye out for privacy and security if you need it to.

## Install

```bash
pip install loggo2
# or
python setup.py install
```

To install with Graylog support, do:
```bash
pip install loggo2[graylog]
```

## Setup

To get started, import and instantiate the main class, ideally somewhere at the core of your project. If you have a module with multiple files, do the initial configuration in the main `__init__.py`, or in a file called `log.py`. so you can import the same, ready-set-up logger easily.

For example, if your app was called `tester`, you could add the following to `tester/__init__.py`:

```python
from loggo2 import Loggo

# all setup values are optional
loggo = Loggo(
    facility="tester",  # name of program logging the message
    graylog_address=("0.0.0.0", 9999),  # address for graylog (ip, port)
    do_print=True,  # print each log to console
    do_write=True,  # write each log to file
    logfile="mylog.txt",  # custom path to logfile
    truncation=1000,  # longest possible value in extra data
    private_data={"password"},  # set of sensitive args/kwargs
)
```

## Usage

In other parts of the project, you can then access the configured logger instance with:

```python
from tester import loggo
```

### Loggo as decorator

You can use `@loggo` as a decorator on any callable: a class, on its method, or on function. On classes, it will log every method; on methods and functions it will log the call signature, return and errors. The central idea behind `loggo` is that you can simply decorate every class in your project, as well as any important standalone functions, and have comprehensive, standardised information about your project's internals without any extra labour.

If a method within a decorated class is called too often, or if you don't need to keep an eye on it, you can use `@loggo.ignore` to ignore it. Also available is `@loggo.errors`, which will only log exceptions, not calls and returns.

For an example use-case, let's make a simple class that multiplies two numbers, but only if a password is supplied. We will ignore logging of the boring authentication system.

```python
@loggo
class Multiplier:
    def __init__(self, base):
        self.base = base

    def multiply(self, n, password):
        """
        Multiply by the number given during initialisation--requires password
        """
        self.authenticated = self._do_authentication(password)
        if not self.authenticated:
            raise ValueError("Not authenticated!")
        return self.base * n

    @loggo.ignore
    def _do_authentication(self, password):
        """Not exactly Fort Knox"""
        return password == "tOpSeCrEt"
```

First, let's use it properly, with our secret password passed in:

```python
mult = Multiplier(50)
result = mult.multiply(50, "tOpSeCrEt")
assert result == 2500  # True
```

We'll get some nice text in the console:

```
11.05 2018 17:14:54 *Called Multiplier.multiply(n=50, password='******')
11.05 2018 17:14:54 *Returned from Multiplier.multiply(n=50, password='******') with int (2500)
```

Notice that our private argument `password` was successfully obscured, even without us naming the argument when we called the method. If you used `do_write=True`, this log will also be in your specified log file, also with password obscured.

```python
result = mult.multiply(7, "password123")
```

Here an error will raise, a log will be generated, and we'll get extra info in the console, including traceback:

```
11.05 2018 17:19:43 *Called Multiplier.multiply(n=7, password='******')
11.05 2018 17:19:43 *Errored during Multiplier.multiply(n=7, password='******') with ValueError "Not authenticated!"    20 -- see below:
Traceback (most recent call last):
  File "/Users/danny/work/loggo/loggo/loggo.py", line 137, in full_decoration
    response = function(*args, **kwargs)
  File "tester.py", line 13, in multiply
    raise ValueError('Not authenticated!')
ValueError: Not authenticated!
```

If you're using [graypy](https://github.com/severb/graypy/), you'll get a lot of extra goodness, such as key-value pairs for call signatures, timestamps, arguments, return values, exception information, and so on.

### Custom messages

When configuring `loggo`, you can use your own message format for the auto-generated logs. There are four keys, one for each autolog type:

```python
loggo = Loggo(
    called="Log before callable is run",
    returned="Log for return from {call_signature} at {timestamp}",
    returned_none="Log when the return value of the callable is None",
    errored="Log string on exception: {exception_type}",
)


@loggo
def test():
    pass
```

If you pass `None` for any of these keyword arguments, logs of that time will be completely suppressed. If you do not provide a value for `returned_none`, `loggo` will use the value you provided for `returned`, or fall back to its own default.

Notice, in the example above, you can include particular format strings in the log message. Currently supported are:
* `call_signature`: the callable name and its arguments and keyword arguments
* `callable`: the `__qualname__` of the decorated object
* `params`: comma separated key value pairs for arguments passed
* `log_level`: the log level associated with this log
* `timestamp`: time at time of logging
* `couplet`: `uuid.uuid1()` for the called and returned/errored pair
* `number_of_params`: total `args + kwargs` as int
* `decorated`: always `True`

The `errored` log additionally supports:
* `exception_type`: `ValueError`, `AttributeError`, etc.
* `exception_msg`: details about the thrown exception
* `traceback`: exception traceback

And the `returned` and `returned_none` logs support:
* `return_value`: the object returned by the callable
* `return_type`: type of returned object

Adding more such strings is trivial; submit an issue if there is something else you need.

### Logging without decorators

For logging manually, `loggo` provides methods similar to the logging functions of the `logging` standard library: `loggo.log`, `loggo.debug`, `loggo.info`, `loggo.warning`, `loggo.error`, and `loggo.critical`. The methods use the configuration that has already been defined. The main method `loggo.log` takes three parameters:

```python
level = 50
msg = "Message to log"
extra = dict(some="data", that="will", be="logged")
loggo.log(level, msg, extra)
# console: 11.05 2018 17:36:24 Message to log  50
# extra_data in log file if `do_print` setting is True
```

Methods `loggo.debug`, `loggo.info`, `loggo.warning`, `loggo.error` and `loggo.critical` are convenience methods for setting the log level. For instance,

```python
loggo.warning("A message", dict(some="data"))
```

is equivalent to

```python
loggo.log(logging.WARNING, "A message", dict(some="data"))
```

where `logging.WARNING` is an integer constant imported from the standard library.

The advantage of using `loggo` for these kinds of logs is that `loggo` will make the extra data more readable and truncate very large strings. More importantly, you also still get whatever extras you've configured, like obfuscation of private data, or writing to console/file.

### Methods

You can also start and stop logging with `loggo.start()` and `loggo.stop()`, at any point in your code, though by default, error logs will still get through. If you want to suppress errors too, you can pass in `allow_errors=False`.

### Context managers

You can suppress logs using a context manager. Errors are allowed here by default too:

```python
with loggo.pause(allow_errors=False):
    do_something()
```

## Contributing

### Style
The style is [Black](https://github.com/psf/black), with the following exceptions and extra strictness:
* Maximum line length is 110

CHANGELOG.md
Changelog
=========

This log should always be updated when doing backwards incompatible changes, resulting in a major version bump. Feel free to add a log for lesser version bumps as well, but for major bumps it's a must.

10.1.3
-----
- Upgraded typing-extensions to be >=4.2.0,<5.0.0
  - Needed for Solana compatibility in index-to-address
- Removing Python < 3.9
- Adding Github Actions checks to run test suite
- Minor fixes to previously failing tests

10.1.2
-----
- Fixed
  - Set as Optional a type initialized to None

10.1.1
-----
- Changed
  - "Returned from" log message is now more concise.

10.1.0
-----
- Changed
    - Set default log level to `INFO`

10.0.0
-----
- Changed
    - Package renamed from `loggo` to `loggo2`

9.0.0
-----
- Added
    - Make Loggo init accept a new optional parameter log_formatter, which specifies a formatter used while writing log records to stdout or a file. Loggo remains backward compatible, as the default log formatter is LocalLogFormatter.
    - Add a new log formatter JsonLogFormatter
- Changed
    - `private_data` kwarg of `Loggo.__init__` no longer accepts `None` value.

8.0.0
-----
- Removed
    - The remaining public data members of a Loggo object: `called`, `returned`, `returned_none` and `errored`
- Fixed
    - Logging calls no longer mutate extra data

7.0.0
-----
- Changed
    - Graypy dependency is not installed by default. To install with Graylog support, install with the `graylog` extra: `pip install loggo[graylog]`
    - Don't allow using @loggo.logme to decorate callables. Use @loggo instead.
    - Make Loggo.write_to_file() a private method.
    - Make Loggo init param `raise_logging_errors` default to True
    - Make loggo.loggo module private
    - Reduce a Loggo object's public data members to minimum. Only `called`, `returned`, `returned_none` and `errored` remain public now.

6.0.0
-----
- Added
    - `called`, `returned`, `returned_none` and `errored` kwargs on instantiation for custom log strings

- Changed
    - do not pass a config dict on instantiation, instead just keyword arguments
    - Graylog address is given to init as tuple named graylog_address instead of having the ip and port in separate arguments.
    - default obscured data string `[PRIVATE_DATA] -> ********`. This string is not configurable anymore.
    - default log level changed from logging.INFO to logging.DEBUG

- Removed
    - loggo.events; its functionality is still available during instantiation
    - line_length config, which was not respected anyway, and not really needed.
    - max_dict_depth config. Now always set to the previous default value 5.
    - error_level config. Now always set to the LOG_LEVEL
    - Loggo.verbose contextmanager
    - Loggo.log_errors contextmanager

5.0.0
-----
- Changed
    - `Loggo.listen_to()` no longer takes positional argument `no_graylog_disable_log`. Instead of this, the key-value-pair `'log_if_graylog_disabled': bool` can be included in the config dict when instantiating Loggo.
    - The `error_alert` kwarg of `Loggo.events()` was renamed to `error_level` and is now expected to be integer.
    - The signature `Loggo.log(self, message: str, alert: Optional[str] = None, extra: Optional[Dict] = None, safe: bool = False)` is now `Loggo.log(self, level: int, msg: str, extra: Optional[Dict] = None, safe: bool = False)`.

4.0.0
-----
- Changed
    - Drop support for Python versions lower than 3.6

3.0.0
-----
- Added
    - Added config option: `truncation (int)`, which determines extra data value max length

- Changed
    - Internals updated to allow auto-logging of static/class methods in uninstantiated classes.
    - Slight reformatting of auto-generated messages (better getting of class name, no `__main__`)

- Removed
    - Loggo.everything (the class decorator), now simply use Loggo

2.0.0
-----
- Changed
    - Major refactor, affecting numerous call signatures

- Removed
    - LoggedException
    - colours


loggo/loggo.py
"""
Loggo: safe and automatable logging
"""

from contextlib import contextmanager
from functools import wraps
import inspect
import logging
import os
import pathlib
import sys
import time
import traceback
from typing import Any, Callable, Dict, Generator, Mapping, Optional, Set, Tuple, Union
import uuid

if sys.version_info < (3, 8):
    from typing_extensions import Literal, TypedDict
else:
    from typing import Literal, TypedDict

# you don't need graylog installed
try:
    import graypy
except ImportError:
    graypy = None


# Strings to be formatted for pre function, post function and error during function
DEFAULT_FORMS = dict(
    called="*Called {call_signature}",
    returned="*Returned from {call_signature} with {return_type} {return_value}",
    returned_none="*Returned None from {call_signature}",
    errored='*Errored during {call_signature} with {exception_type} "{exception_msg}"',
)
LOG_LEVEL = logging.DEBUG  # Log level used for Loggo decoration logs
LOG_THRESHOLD = logging.DEBUG  # Only log when log level is this or higher
MAX_DICT_OBSCURATION_DEPTH = 5
OBSCURED_STRING = "********"
# Callables with an attribute of this name set to True will not be logged by Loggo
NO_LOGS_ATTR_NAME = "_do_not_log_this_callable"
DATE_FORMAT = "%Y-%m-%d %H:%M:%S %Z"

# Make a dummy logging.LogRecord object, so that we can inspect what
# attributes instances of that class have.
_dummy_log_record = logging.LogRecord("dummy_name", logging.INFO, "dummy_pathname", 1, "dummy_msg", {}, None)
LOG_RECORD_ATTRS = vars(_dummy_log_record).keys()


class _Formatters(TypedDict, total=False):
    """A dictionary of data that can be input into log messages.

    The keys can be used in log message forms e.g. "*Called
    {call_signature}". In the final log message {call_signature} will
    then be replaced by its value in this formatter dict.
    """

    call_signature: str
    callable: str
    params: str

    decorated: bool
    couplet: uuid.UUID
    number_of_params: int
    private_keys: str
    timestamp: str
    log_level: int

    # Only available if 'errored'
    traceback: str
    exception_type: str
    exception_msg: str

    # Only available if 'returned' or 'returned_none'
    return_value: str
    return_type: str


_CallableEvent = Literal["called", "errored", "returned", "returned_none"]


class _LocalLogFormatter(logging.Formatter):
    """Formatter for file logs and stdout logs."""

    def __init__(self) -> None:
        super().__init__("%(asctime)s\t%(message)s\t%(levelno)s", DATE_FORMAT)

    def format(self, record: logging.LogRecord) -> str:  # noqa: A003
        msg = super().format(record)
        traceback = getattr(record, "traceback", None)
        if traceback:
            msg += " -- see below:\n" + traceback.rstrip("\n")
        return msg


class Loggo:
    """A class for logging."""

    def __init__(
        self,
        *,  # Reject positional arguments
        called: Optional[str] = DEFAULT_FORMS["called"],
        returned: Optional[str] = DEFAULT_FORMS["returned"],
        returned_none: Optional[str] = DEFAULT_FORMS["returned_none"],
        errored: Optional[str] = DEFAULT_FORMS["errored"],
        facility: str = "loggo",
        graylog_address: Optional[Tuple[str, int]] = None,
        do_print: bool = False,
        do_write: bool = False,
        truncation: int = 7500,
        raise_logging_errors: bool = True,
        logfile: str = "./logs/logs.txt",
        private_data: Optional[Set[str]] = None,
        log_if_graylog_disabled: bool = True,
    ) -> None:
        """Initializes a Loggo object.

        On instantiation, pass in a dictionary containing the config.
        Currently accepted config values are:
        - facility: name of the app the log is coming from
        - graylog_address: A tuple (ip, port). Address for graylog.
        - logfile: path to a file to which logs will be written
        - do_print: print logs to console
        - do_write: write logs to file
        - truncation: truncate value of log data fields to this length
        - private_data: key names that should be filtered out of logging. when not
        - raise_logging_errors: should stdlib `log` call errors be suppressed or no?
        - log_if_graylog_disabled: boolean value, should a warning log be made when failing to
            connect to graylog
        """
        self.stopped = False
        self.allow_errors = True
        self.called = called
        self.returned = returned
        self.returned_none = self._best_returned_none(returned, returned_none)
        self.errored = errored
        self.facility = facility
        self.graylog_address = graylog_address
        self.do_print = do_print
        self.do_write = do_write
        self.truncation = truncation
        self.raise_logging_errors = raise_logging_errors
        self.logfile = os.path.abspath(os.path.expanduser(logfile))
        self.private_data = private_data or set()
        self.log_if_graylog_disabled = log_if_graylog_disabled
        self.logger = logging.getLogger(self.facility)
        self.logger.setLevel(LOG_THRESHOLD)

        if do_write:
            # create the directory where logs are stored if it does not exist yet
            pathlib.Path(os.path.dirname(self.logfile)).mkdir(parents=True, exist_ok=True)
            file_handler = logging.FileHandler(self.logfile, delay=True)
            file_handler.setFormatter(_LocalLogFormatter())
            self.logger.addHandler(file_handler)

        if do_print:
            print_handler = logging.StreamHandler(sys.stdout)
            print_handler.setFormatter(_LocalLogFormatter())
            self.logger.addHandler(print_handler)

        self._add_graylog_handler()

    def __call__(self, class_or_func: Union[Callable, type]) -> Union[Callable, type]:
        """Make Loggo object itself a decorator.

        Allow decorating either a class or a method/function, so @loggo
        can be used on both classes and functions.
        """
        if isinstance(class_or_func, type):
            return self._decorate_all_methods(class_or_func)
        if self._can_decorate(class_or_func):
            return self._logme(class_or_func)
        return class_or_func

    @staticmethod
    def _get_timestamp() -> str:
        """Return current time as a string.

        Formatted as follows: "2019-07-17 09:35:06 CEST".
        """
        return time.strftime(DATE_FORMAT, time.localtime())

    @staticmethod
    def _best_returned_none(returned: Optional[str], returned_none: Optional[str]) -> Optional[str]:
        """Resolve the format for a log message, when a function returns None.

        If the user has their own msg format for 'returned' logs, but
        not one for 'returned_none', we should use theirs over loggo's
        default.
        """
        # if the user explicitly doesn't want logs for returns, set to none
        if not returned_none or not returned:
            return None
        # if they provided their own, use that
        if returned_none != DEFAULT_FORMS["returned_none"]:
            return returned_none
        # if the user just used the defaults, use those
        if returned == DEFAULT_FORMS["returned"]:
            return returned_none
        # the switch: use the user provided returned for returned_none
        return returned

    @staticmethod
    def _can_decorate(candidate: Callable, name: Optional[str] = None) -> bool:
        """Decide if we can decorate a given callable.

        Don't decorate python magic methods.
        """
        name = name or getattr(candidate, "__name__", None)
        if not name:
            return False
        if name.startswith("__") and name.endswith("__"):
            return False
        return True

    def _decorate_all_methods(self, cls: type, just_errors: bool = False) -> type:
        """Decorate all viable methods in a class."""
        members = inspect.getmembers(cls)
        members = [(k, v) for k, v in members if callable(v) and self._can_decorate(v, name=k)]
        for name, candidate in members:
            deco = self._logme(candidate, just_errors=just_errors)
            # somehow, decorating classmethods as staticmethods is the only way
            # to make everything work properly. we should find out why, some day
            if isinstance(vars(cls)[name], (staticmethod, classmethod)):
                # Make mypy ignore due to an open issue: https://github.com/python/mypy/issues/5530
                deco = staticmethod(deco)  # type: ignore
            try:
                setattr(cls, name, deco)
            # AttributeError happens if we can't write, as with __dict__
            except AttributeError:
                pass
        return cls

    @contextmanager
    def pause(self, allow_errors: bool = True) -> Generator[None, None, None]:
        """A context manager that prevents loggo from logging in that context.

        By default, errors will still make it through, unless
        allow_errors==False
        """
        original = self.allow_errors, self.stopped
        self.stopped = True
        self.allow_errors = allow_errors
        try:
            yield
        finally:
            self.allow_errors, self.stopped = original

    def stop(self, allow_errors: bool = True) -> None:
        """Stop loggo from logging.

        By default still log raised exceptions.
        """
        self.stopped = True
        self.allow_errors = allow_errors

    def start(self, allow_errors: bool = True) -> None:
        """Continue logging after a call to `stop` or inside a `pause`."""
        self.stopped = False
        self.allow_errors = allow_errors

    @staticmethod
    def ignore(function: Callable) -> Callable:
        """A decorator that will override Loggo class decorator.

        If a class is decorated with @loggo, logging can still be
        disabled for certain methods using this decorator.
        """
        setattr(function, NO_LOGS_ATTR_NAME, True)
        return function

    def errors(self, class_or_func: Union[Callable, type]) -> Union[Callable, type]:
        """
        Decorator: only log errors within a given method
        """
        if isinstance(class_or_func, type):
            return self._decorate_all_methods(class_or_func, just_errors=True)
        return self._logme(class_or_func, just_errors=True)

    def _logme(self, function: Callable, just_errors: bool = False) -> Callable:
        """A decorator for automated input/output logging.

        Used by @loggo and @loggo.errors decorators. Makes a log when a
        callable is called, returns, or raises. If `just_errors` is
        True, only logs when a callable raises.
        """
        # if logging has been turned off, just do nothing
        if getattr(function, NO_LOGS_ATTR_NAME, False):
            return function

        @wraps(function)
        def full_decoration(*args: Any, **kwargs: Any) -> Any:
            """Main decorator logic.

            Generate a log before running the callable, then try to run
            it. If it errors, log the error. If it doesn't, log the
            return value.
            """
            bound = self._params_to_dict(function, *args, **kwargs)
            if bound is None:
                self.warning(
                    "Failed getting function signature, or coupling arguments with signature's parameters",
                    extra={"callable_name": getattr(function, "__qualname__", "unknown_callable")},
                )
                return function(*args, **kwargs)

            param_strings = self.sanitise(bound)
            formatters = self._make_call_signature(function, param_strings)
            privates = [key for key in param_strings if key not in bound]

            # add more format strings
            more = _Formatters(
                decorated=True,
                couplet=uuid.uuid1(),
                number_of_params=len(args) + len(kwargs),
                private_keys=", ".join(privates),
                timestamp=self._get_timestamp(),
            )
            formatters.update(more)

            # 'called' log tells you what was called and with what arguments
            if not just_errors:
                self._generate_log("called", None, formatters, param_strings)

            try:
                # where the original function is actually run
                response = function(*args, **kwargs)
            # handle any possible error in the original function
            except Exception as error:
                formatters["traceback"] = traceback.format_exc()
                self._generate_log("errored", error, formatters, param_strings)
                raise
            where: _CallableEvent = "returned_none" if response is None else "returned"
            # the successful return log
            if not just_errors:
                self._generate_log(where, response, formatters, param_strings)
            # return whatever the original callable did
            return response

        return full_decoration

    def _string_params(self, non_private_params: Dict, use_repr: bool = True) -> Dict[str, str]:
        """Turn every entry in log_data into truncated strings."""
        params = dict()
        for key, val in non_private_params.items():
            truncation = self.truncation if key not in {"trace", "traceback"} else None
            safe_key = self._force_string_and_truncate(key, 50, use_repr=False)
            safe_val = self._force_string_and_truncate(val, truncation, use_repr=use_repr)
            params[safe_key] = safe_val
        return params

    @staticmethod
    def _make_call_signature(function: Callable, param_strings: Dict[str, str]) -> _Formatters:
        """Represent the call as a string mimicking how it is written in
        Python.

        Return it within a dict containing some other format strings.
        """
        signature = "{callable}({params})"
        param_str = ", ".join(f"{k}={v}" for k, v in param_strings.items())
        format_strings = _Formatters(
            callable=getattr(function, "__qualname__", "unknown_callable"), params=param_str
        )
        format_strings["call_signature"] = signature.format(**format_strings)
        return format_strings

    def listen_to(loggo_self, facility: str) -> None:
        """Listen to logs from another logger and make loggo log them.

        This method can hook the logger up to anything else that logs
        using the Python logging module (i.e. another logger) and steals
        its logs. This can be useful for instance for logging logs of a
        library using a shared Loggo configuration.
        """

        class LoggoHandler(logging.Handler):
            def emit(handler_self, record: logging.LogRecord) -> None:
                extra = {k: v for k, v in vars(record).items() if k not in LOG_RECORD_ATTRS}
                extra["sublogger"] = facility
                loggo_self.log(record.levelno, record.msg, extra)

        other_logger = logging.getLogger(facility)
        other_logger.setLevel(LOG_THRESHOLD)
        other_logger.addHandler(LoggoHandler())

    @staticmethod
    def _params_to_dict(function: Callable, *args: Any, **kwargs: Any) -> Optional[Mapping]:
        """Turn args and kwargs into an OrderedDict of {param_name: value}.

        Returns None if getting the signature, or binding arguments to
        the signature's parameters fails.
        """
        try:
            sig = inspect.signature(function)
        except ValueError:
            return None

        try:
            bound_obj = sig.bind(*args, **kwargs)
        except TypeError:
            return None

        bound = bound_obj.arguments
        if bound:
            first = list(bound)[0]
            if first == "self":
                bound.pop("self")
            elif first == "cls":
                bound.pop("cls")
        return bound

    def _obscure_private_keys(self, log_data: Any, dict_depth: int = 0) -> Any:
        """Obscure any private values in a dictionary recursively."""
        if not isinstance(log_data, dict) or dict_depth >= MAX_DICT_OBSCURATION_DEPTH:
            return log_data

        out = dict()
        for key, value in log_data.items():
            if key in self.private_data:
                out[key] = OBSCURED_STRING
            else:
                out[key] = self._obscure_private_keys(value, dict_depth + 1)
        return out

    def _represent_return_value(self, response: Any) -> str:
        """Make a string representation of whatever a method returns."""
        # some custom handling for request response objects
        if str(type(response)) == "<class 'requests.models.Response'>":
            response = response.text

        return "({})".format(self._force_string_and_truncate(response, truncate=None, use_repr=True))

    def _generate_log(
        self, where: _CallableEvent, returned: Any, formatters: _Formatters, safe_log_data: Dict[str, str]
    ) -> None:
        """Generate message, level and log data for automated logs.

        - msg (str): the unformatted message
        - returned (ANY): what the decorated callable returned
        - formatters (dict): dict containing format strings needed for message
        - safe_log_data (dict): dict of stringified, truncated, censored parameters
        """
        # if the user turned off logs of this type, do nothing immediately
        msg = getattr(self, where)
        if not msg:
            return

        # if errors not to be shown and this is an error, quit
        if not self.allow_errors and where == "errored":
            return

        # if state is stopped and not an error, quit
        if self.stopped and where != "errored":
            return

        # do not log loggo, because why would you ever want that?
        if "loggo.loggo" in formatters["call_signature"]:
            return

        # return value for log message
        if where in {"returned", "returned_none"}:
            ret_str = self._represent_return_value(returned)
            formatters["return_value"] = ret_str
            formatters["return_type"] = type(returned).__name__

        # if what is 'returned' is an exception, get the error formatters
        if where == "errored":
            formatters["exception_type"] = type(returned).__name__
            formatters["exception_msg"] = str(returned)
        formatters["log_level"] = LOG_LEVEL

        # format the string template
        msg = msg.format(**formatters).replace("  ", " ")

        # make the log data
        log_data = {**formatters, **safe_log_data}
        custom_log_data = self.add_custom_log_data()
        log_data.update(custom_log_data)

        # record if logging was on or off
        original_state = self.stopped
        # turn it on just for now, as if we shouldn't log we'd have returned
        self.stopped = False
        try:
            self.log(LOG_LEVEL, msg, extra=log_data, safe=True)
        finally:
            # restore old stopped state
            self.stopped = original_state

    def add_custom_log_data(self) -> Dict[str, str]:
        """An overwritable method useful for adding custom log data."""
        return dict()

    def _add_graylog_handler(self) -> None:
        if not graypy:
            if self.graylog_address:
                raise ValueError("Misconfiguration: Graylog configured but graypy not installed")
            return

        if not self.graylog_address:
            if self.log_if_graylog_disabled:
                self.warning("Graypy installed, but Graylog not configured! Disabling it")
            return

        handler = graypy.GELFUDPHandler(*self.graylog_address, debugging_fields=False)
        self.logger.addHandler(handler)

    def _force_string_and_truncate(self, obj: Any, truncate: Optional[int], use_repr: bool = False) -> str:
        """Return stringified and truncated obj.

        If stringification fails, log a warning and return the string
        '<<Unstringable input>>'
        """
        try:
            obj = str(obj) if not use_repr else repr(obj)
        except Exception as exc:
            self.warning(
                "Object could not be cast to string", extra=dict(exception_type=type(exc), exception=exc)
            )
            return "<<Unstringable input>>"
        if truncate is None:
            return obj
        # truncate and return
        return (obj[:truncate] + "...") if len(obj) > (truncate + 3) else obj

    @staticmethod
    def _rename_protected_keys(log_data: Dict) -> Dict:
        """Rename log data keys with valid names.

        Some names cannot go into logger. Rename the invalid keys with a
        prefix before logging.
        """
        out = dict()
        # Names that stdlib logger will not like. Based on [1]
        # [1]: https://github.com/python/cpython/blob/04c79d6088a22d467f04dbe438050c26de22fa85/Lib/logging/__init__.py#L1550  # noqa: E501
        protected = {"message", "asctime"} | LOG_RECORD_ATTRS
        for key, value in log_data.items():
            if key in protected:
                key = "protected_" + key
            out[key] = value
        return out

    def sanitise(self, unsafe_dict: Mapping, use_repr: bool = True) -> Dict[str, str]:
        """Ensure that log data is safe to log.

        - No private keys
        - Rename protected keys
        - Everything strings
        """
        obscured = self._obscure_private_keys(unsafe_dict)
        no_protected = self._rename_protected_keys(obscured)
        return self._string_params(no_protected, use_repr=use_repr)

    def sanitise_msg(self, msg: str) -> str:
        """Overwritable method to clean or alter log messages."""
        return msg

    def log(self, level: int, msg: str, extra: dict = None, safe: bool = False) -> None:
        """Main logging method, called both in auto logs and manually by user.

        level: int, priority of log
        msg: string to log
        extra: dict of extra fields to log
        safe: do we need to sanitise extra?
        """
        # don't log in a stopped state
        if self.stopped:
            return

        extra = extra or dict()

        if not safe:
            extra = self.sanitise(extra, use_repr=False)
            msg = self.sanitise_msg(msg)

        extra.update(dict(log_level=str(level), loggo="True"))

        try:
            self.logger.log(level, msg, extra=extra)
        # The log call shouldn't ever fail, because of the way we rename protected
        # keys in `extra`. For the paranoid, we still keep the option to swallow
        # any unexpected errors (due to possible bugs in a 3rd party Handler etc.).
        except Exception:
            if self.raise_logging_errors:
                raise

    def debug(self, *args: Any, **kwargs: Any) -> None:
        return self.log(logging.DEBUG, *args, **kwargs)

    def info(self, *args: Any, **kwargs: Any) -> None:
        return self.log(logging.INFO, *args, **kwargs)

    def warning(self, *args: Any, **kwargs: Any) -> None:
        return self.log(logging.WARNING, *args, **kwargs)

    def error(self, *args: Any, **kwargs: Any) -> None:
        return self.log(logging.ERROR, *args, **kwargs)

    def critical(self, *args: Any, **kwargs: Any) -> None:
        return self.log(logging.CRITICAL, *args, **kwargs)
